code/modules/mob/living/silicon/robot/robot.dm

```diff
@@ -158,6 +160,19 @@
 		rbPDA = new/obj/item/device/pda/ai(src)
 	rbPDA.set_name_and_job(custom_name,braintype)
 
+/mob/living/silicon/robot/proc/upgrade_components()
+	if(component_extension)
+		for(var/V in components) if(V != "power cell")
+			var/datum/robot_component/C = components[V]
+			var/NC = text2path("[C.external_type][component_extension]")
+			var/obj/item/robot_parts/robot_component/I = new NC
```

Comment:
You can assign `NC` to this, instead of `new NC`
Then, you can abuse `initial()` to avoid creating and deleting objects that you don't need.